src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -0,0 +1,258 @@
+            if (requestHandle.IsInvalid)
+            {
+                throw new HttpRequestException(SR.net_http_client_execution_error);
+            }
+
+            Interop.libcurl.curl_easy_setopt(requestHandle, CURLoption.CURLOPT_URL, state.RequestMessage.RequestUri.OriginalString);
+            if (_automaticRedirection)
+            {
+                Interop.libcurl.curl_easy_setopt(requestHandle, CURLoption.CURLOPT_FOLLOWLOCATION, 1);
+            }
+
+            // TODO: Handle headers, proxy and other options
+
+            return requestHandle;
+        }
+
+
+        private HttpResponseMessage CreateResponseMessage(SafeCurlHandle requestHandle, HttpRequestMessage request)
+        {
+            var response = new HttpResponseMessage();
+            Int64 httpStatusCode = 0;
+            int result = Interop.libcurl.curl_easy_getinfo(requestHandle, CURLINFO.CURLINFO_RESPONSE_CODE, ref httpStatusCode);
+            if (CURLcode.CURLE_OK != result)
+            {
+                throw new HttpRequestException(SR.net_http_client_execution_error, GetCurlException(result));
+            }
+            response.StatusCode = (HttpStatusCode)httpStatusCode;
+
+            // TODO: Do error processing if needed and return actual response
+            response.Content = new StringContent("SendAsync to " + request.RequestUri.OriginalString + " returned: " + response.StatusCode);
+            return response;
+        }
+
+        private void HandleAsyncException(RequestState state, Exception ex)
+        {
+            if (state.CancellationToken.IsCancellationRequested)
+            {
+                // If the exception was due to the cancellation token being canceled, throw cancellation exception.
+                state.Tcs.TrySetCanceled();
+            }
+            else if (ex is HttpRequestException)
+            {
+                state.Tcs.TrySetException(ex);
+            }
+            else
+            {
+                state.Tcs.TrySetException(new HttpRequestException(SR.net_http_client_execution_error, ex));
+            }
+        }
+
+        private void CheckDisposed()
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(GetType().FullName);
+            }
+        }
+
+        private void CheckDisposedOrStarted()
+        {
+            CheckDisposed();
+            if (_anyOperationStarted)
+            {
+                throw new InvalidOperationException(SR.net_http_operation_started);
+            }
+        }
+
+        private string GetCurlErrorString(int code)
+        {
+            IntPtr ptr = Interop.libcurl.curl_easy_strerror(code);
+            return Marshal.PtrToStringAnsi(ptr);
+        }
+
+        private Exception GetCurlException(int code)
+        {
+            return new Exception(GetCurlErrorString(code));
+        }
+
+        #endregion
+
+        private class RequestState
```

Comment:
The same comment @stephentoub made for WinHttpHandler applies here:

> As we're looking at perf later, there are a variety of improvements that can be made here. For example, it looks like RequestState could derive from TaskCompletionSource instead of having a field of one, and it could have a field to store this. That has a few benefits. First, it shrinks the size of the combination of RequestSize + TaskCompletionSource by several fields (state.Tcs, the extra object header, etc.) Second, it means that the state passed to StartNew already contains the this reference necessary to call the instance StartRequest method... that lets you rewrite the call to use a lambda like s => ((RequestState)s).thisRef.StartRequest(), which in turn lets the compiler cache the delegate and avoid the delegate allocation this currently incurs.

https://github.com/dotnet/corefx/pull/2149#discussion_r33181993